weather_dl/download_pipeline/manifest.py

```diff
@@ -559,6 +566,116 @@ def _update(self, download_status: DownloadStatus) -> None:
             logger.debug(download_status)
 
 
+def get_wait_interval(num_retries: int = 0) -> float:
+    """Returns next wait interval in seconds, using an exponential backoff algorithm."""
+    if 0 == num_retries:
+        return 0
+    return 2 ** num_retries
+
+
+class FirestoreManifest(Manifest):
+    """A Firestore Manifest.
+    This Manifest implementation stores DownloadStatuses in a Firebase document store.
+    The document hierarchy for the manifest is as follows:
+      [manifest  <or manifest name, configurable from CLI>]
+      ├── doc_id (md5 hash of the path) { 'selection': {...}, 'location': ..., 'username': ... }
+      └── etc...
+    Where `[<name>]` indicates a collection and `<name> {...}` indicates a document.
+    """
+
+    def _get_db(self) -> firestore.firestore.Client:
+        """Acquire a firestore client, initializing the firebase app if necessary.
+        Will attempt to get the db client five times. If it's still unsuccessful, a
+        `ManifestException` will be raised.
+        """
+        db = None
+        attempts = 0
+
+        while db is None:
+            try:
+                db = firestore.client()
+            except ValueError as e:
+                # The above call will fail with a value error when the firebase app is not initialized.
+                # Initialize the app here, and try again.
+                firebase_admin.initialize_app(options=self.get_firestore_config())
+                logger.info('Initialized Firebase App.')
+
+                if attempts > 4:
+                    raise ManifestException('Exceeded number of retries to get firestore client.') from e
+
+            time.sleep(get_wait_interval(attempts))
+
+            attempts += 1
+
+        return db
+
+    def _read(self, location: str) -> DownloadStatus:
+        """Reads the JSON data from a manifest."""
+
+        doc_id = hashlib.md5(location.encode('utf-8')).hexdigest()
```

Comment:
Might be nice to have a method for this.

Reply:
Done.